packages/pubsub/src/publisher.js

```diff
@@ -0,0 +1,179 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/publisher
+ */
+
+'use strict';
+
+var arrify = require('arrify');
+var common = require('@google-cloud/common');
+var extend = require('extend');
+var is = require('is');
+
+/**
+ *
+ */
+function Publisher(topic, options) {
+  this.topic = topic;
+  this.api = topic.api;
+
+  this.inventory = {
+    queued: [],
+    inFlight: []
+  };
+
+  Object.defineProperty(this.inventory, 'queueSize', {
+    get: function() {
+      return this.queued.reduce(function(size, message) {
+        return size + message.data.length;
+      }, 0);
+    }
+  });
+
+  this.settings = extend(true, {
+    batching: {
+      maxBytes: Infinity,
+      maxMessages: Infinity,
+      maxMilliseconds: 1000
+    }
+  }, options);
+
+  this.publishHandle_ = null;
+}
+
+/**
+ *
+ */
+Publisher.prototype.cancelPublish_ = function() {
+  clearTimeout(this.publishHandle_);
+  this.publishHandle_ = null;
+};
+
+/**
+ *
+ */
+Publisher.prototype.prepareMessages_ = function() {
+  var size = 0;
+  var messages = [];
+
+  var message;
+
+  while (message = this.inventory.queued.shift()) {
+    var isUnderSizeLimit = (size + message.data.length) <=
+      this.settings.batching.maxBytes;
+
+    var isUnderMessageLimit = (messages.length + 1) <=
+      this.settings.batching.maxMessages;
+
+    if (!isUnderSizeLimit || !isUnderMessageLimit) {
+      this.inventory.queued.unshift(message);
+      break;
+    }
+
+    messages.push(message);
+  }
+
+  return messages;
+};
+
+/**
+ *
+ */
+Publisher.prototype.publish = function(data, attrs, callback) {
+  if (!(data instanceof Buffer)) {
+    throw new Error('Data must be in the form of a Buffer.');
+  }
+
+  if (is.fn(attrs)) {
+    callback = attrs;
+    attrs = {};
+  }
+
+  this.inventory.queued.push({
+    data: data,
+    attrs: attrs,
+    callback: callback
+  });
+
+  var reachedMaxMessages = this.inventory.queued.length >=
+    this.settings.batching.maxMessages;
+
+  var reachedMaxBytes = this.inventory.queueSize >=
+    this.settings.batching.maxBytes;
+
+  // publish immediately if we hit maxMessage or maxBytes
+  if (reachedMaxMessages || reachedMaxBytes) {
+    this.publish_();
+  } else if (!this.timeoutHandle_) {
+    this.schedulePublish_();
+  }
+};
+
+/**
+ *
+ */
+Publisher.prototype.publish_ = function() {
+  var self = this;
+
+  this.cancelPublish_();
+
+  var messages = this.prepareMessages_();
+
+  var reqOpts = {
+    topic: this.topic.name,
+    messages: messages.map(function(message) {
+      return {
+        data: message.data,
+        attributes: message.attrs
+      };
+    })
+  };
+
+  this.inventory.inFlight = messages;
```

Comment:
So, the addition of `maxRequests` missed the point of the issue I was raising before. The issue I was raising is: I think it is a problem that the single batch is trying to juggle its messages in the way that you are doing. Adding a `maxRequests` parameter does nothing to solve this (and creates new problems).

Similarly, when I said I thought it was okay to have a few concurrent requests, I did not mean "add a cap on the number of concurrent requests"; I meant the exact opposite, which was _you do not need to worry about this problem_.

The thing I want to make sure of is that multiple batches can be in flight at once, because that will happen. I want to make sure that case works. I do not want to worry about arbitrarily capping how many batches can be in flight; I do not see this as a serious problem.